### ISSUE TITLE
running doParallel with caret 6.0-77 and keras

### ISSUE THREAD

**Issue by sparcycram** [topepo/caret#750]
I am struggling with the following. It appears as though the latest version of caret does not work with do Parallel?

I am also not able to run keras mlps

The errors I receive are:-

Error in e$fun(obj, substitute(ex), parent.frame(), e$data) : 
  unable to find variable "optimismBoot"

Error: Could not start a sequential model. `tensorflow` might not be installed. See `?install_tensorflow`.
In addition: Warning message:
In stop(e, call. = FALSE) : additional arguments ignored in stop()

# Code below

library(caret)
set.seed(1)
dat <- twoClassSim(100)
X <- dat[,1:5]
y <- dat[["Class"]]

library(doParallel)
cl <- makeCluster(detectCores(), type = 'PSOCK')
registerDoParallel(cl)

cl

fitControl <- trainControl(## 10-fold CV
  method = "repeatedcv",
  number = 10,
  repeats = 10)

glmnetfit <- train(X, y,
                   trControl = fitControl,
                   method = "glmnet")

glmnetfit

ptm_A <- proc.time()
library("keras")
library("tensorflow")
mlpKerasDropoutfit <- train(X, y,
                            trControl = fitControl,
                            method = "mlpKerasDropout")

mlpKerasDropoutfit

> sessionInfo()
R version 3.4.2 (2017-09-28)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

Matrix products: default

locale:
[1] LC_COLLATE=English_United Kingdom.1252  LC_CTYPE=English_United Kingdom.1252    LC_MONETARY=English_United Kingdom.1252
[4] LC_NUMERIC=C                            LC_TIME=English_United Kingdom.1252    

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] tensorflow_1.4    keras_2.0.8       doParallel_1.0.11 iterators_1.0.8   foreach_1.4.3     caret_6.0-77      ggplot2_2.2.1    
[8] lattice_0.20-35  

loaded via a namespace (and not attached):
 [1] reticulate_1.2     kernlab_0.9-25     reshape2_1.4.2     purrr_0.2.3        splines_3.4.2      colorspace_1.3-2   stats4_3.4.2      
 [8] base64enc_0.1-3    survival_2.41-3    prodlim_1.6.1      rlang_0.1.2        ModelMetrics_1.1.0 e1071_1.6-8        glue_1.1.1        
[15] withr_2.0.0        bindrcpp_0.2       bindr_0.1          plyr_1.8.4         dimRed_0.1.0       lava_1.5.1         robustbase_0.92-7 
[22] stringr_1.2.0      timeDate_3012.100  munsell_0.4.3      gtable_0.2.0       recipes_0.1.0      codetools_0.2-15   tfruns_1.0        
[29] class_7.3-14       DEoptimR_1.0-8     Rcpp_0.12.13       scales_0.5.0       ipred_0.9-6        CVST_0.2-1         jsonlite_1.5      
[36] stringi_1.1.5      dplyr_0.7.4        RcppRoll_0.2.2     ddalpha_1.3.1      grid_3.4.2         tools_3.4.2        magrittr_1.5      
[43] lazyeval_0.2.0     glmnet_2.0-13      tibble_1.3.4       whisker_0.3-2      DRR_0.0.2          pkgconfig_2.0.1    MASS_7.3-47       
[50] Matrix_1.2-11      lubridate_1.6.0    gower_0.1.2        assertthat_0.2.0   R6_2.2.2           rpart_4.1-11       sfsmisc_1.1-1     
[57] nnet_7.3-12        nlme_3.1-131       compiler_3.4.2    
> 

Any help appreciated

**Comment by coforfe** [comment]
It is not  supported to be run in parallel according to the documentation:

[http://topepo.github.io/caret/train-models-by-tag.html#Neural_Network](here)

```{r}
*Multilayer Perceptron Network with Dropout*

  method = 'mlpKerasDropout'
Type: Regression, Classification

Tuning parameters:

size (#Hidden Units)
dropout (Dropout Rate)
batch_size (Batch Size)
lr (Learning Rate)
rho (Rho)
decay (Learning Rate Decay)
activation (Activation Function)
Required packages: keras

Notes: After train completes, the keras model object is serialized so that it can be used between R session. 
When predicting, the code will temporarily unsearalize the object. 
To make the predictions more efficient, the user might want to use keras::unsearlize_model(object$finalModel$object) in the current R session so that that operation is only done once. 
**Also, this model cannot be run in parallel due to the nature of how tensorflow does the computations**.

```

Thanks,
Carlos.

**Comment by topepo** [comment]
You should also make sure that `install_tensorflow` was run successfully.

**Comment by KafeelBasha** [comment]
Why do we need the parameter 'rho' in 'mlpKerasDropout' method?
